### PR TITLE
fix: fix show,ProcessingStep (issue #162)

### PR DESCRIPTION
**Pull request by jorainer**
This fixes issue #162 (as well as issue #70)

### CODE REVIEW

**Comment by lgatto** [comment]
Here's a real-life example that isn't very readable. I don't think printing de function body is a good solution; `show` print a much shorter and readable, I think.

```r
> x <- Spectra(fl)
> ## function used as arguement to filterIntensity()
> low_int <- function(x, percentage = 0.05) {
+   x > max(x, na.rm = TRUE) * percentage
+ }
> x1 <- filterRt(x, rt = c(rtime(x)[1] - 24, rtime(x)[1] + 24))
> x1 <- filterEmptySpectra(x1)
> x1 <- filterMzRange(x1, c(1648, 1655))
> x1 <- combineSpectra(x1, intensityFun = sum)
Backend of the input object is read-only, will change that to an 'MsBackendDataFrame'
> x1 <- filterIntensity(x1, low_int)
> x1 <- pickPeaks(x1)
> x1@processingQueue
[[1]]
Object of class "ProcessingStep"
 Function: function(x, spectrumMsLevel, intensity,
                                             msLevel = spectrumMsLevel, ...) {



    if (!spectrumMsLevel %in% msLevel || !length(x))
        return(x)
    keep <- intensity(x[, "intensity"])
    if (!is.logical(keep) || length(keep) != nrow(x))
        stop("Error in filterIntensity: the provided function does not return ",
             "a logical vector of length equal to the number of peaks.",
             call. = FALSE)
    x[which(keep), , drop = FALSE]
}
<environment: namespace:Spectra>
 Arguments:
  o  intensity  = function(x, percentage = 0.05) {
  x > max(x, na.rm = TRUE) * percentage
}
  o  msLevel  = [1] 1

[[2]]
Object of class "ProcessingStep"
 Function: function(x, spectrumMsLevel, centroided = NA,
                        halfWindowSize = 2L, method = c("MAD", "SuperSmoother"),
                        snr = 0L, k = 0L, descending = FALSE, threshold = 0,
                        msLevel = spectrumMsLevel, ...) {
    if (!(spectrumMsLevel %in% msLevel))
        return(x)
    if (!nrow(x)) {
        warning("Spectrum is empty. Nothing to pick.")
        return(x)
    }

    n <- noise(x[, 1L], x[, 2L], method = method, ...)

    l <- localMaxima(x[, 2L], hws = halfWindowSize)

    p <- which(l & x[, 2L] > (snr * n))

    if (k > 0L) {
        cbind(mz = refineCentroids(x = x[, 1L], y = x[, 2L], p = p,
                                   k = k, threshold = threshold,
                                   descending = descending),
              intensity = x[p, 2L])
    } else {
        x[p, , drop = FALSE]
    }
}
<environment: namespace:Spectra>
 Arguments:
  o  halfWindowSize  = [1] 2
  o  method  = [1] "MAD"
  o  snr  = [1] 0
  o  k  = [1] 0
  o  descending  = [1] FALSE
  o  threshold  = [1] 0
  o  msLevel  = [1] 1
```

**Review by lgatto — COMMENTED** [review]
See comment in the PR conversation.

**Comment by jorainer** [comment]
Agree, that doesn't look good at all. I first tried to just get the function name with `deparse(substitute(` but that did not work for the arguments - there it was then simply printing `args[1]`.
So maybe your solution is better. I will change.

**Comment by lgatto** [comment]
Maybe an exception, but for anonymous functions, there wouldn't be any name at all.

**Comment by jorainer** [comment]
`show,ProcessingStep` will now print `"user-provided function"` if an argument or `@FUN` is a `function`.

File: tests/testthat/test_ProcessingStep.R [hunk 1]
```diff
@@ -15,3 +15,20 @@ test_that("ProcessingStep executeProcessingStep", {
...
+test_that(".cat_fun works", {
+    res <- .cat_fun("sum")
+    expect_equal(res, "sum")
+    res <- .cat_fun(sum)
+    expect_equal(res, "user-provided function")
+})
+
+test_that("show,ProcessingStep works", {
+    ps <- ProcessingStep(sum, list(1:4))
+    show(ps)
```
Comment (sgibb) [inline tests/testthat/test_ProcessingStep.R:28]: That's not a real test. This just fails if there is an error. What about `expect_output`?

**Review by sgibb — CHANGES_REQUESTED** [review]
Nice fix but the unit tests for `show` don't really test anything.

File: tests/testthat/test_ProcessingStep.R (hunk 1 above)
Comment (jorainer) [reply tests/testthat/test_ProcessingStep.R:28]: :) you got me - I was just too lazy... I will fix.

**Comment by lgatto** [comment]
I now get
```r
> x1@processingQueue
[[1]]
Object of class "ProcessingStep"
 Function: user-provided function
 Arguments:
  o intensity = user-provided function  o msLevel = 1
[[2]]
Object of class "ProcessingStep"
 Function: user-provided function
 Arguments:
  o halfWindowSize = 2  o method = MAD  o snr = 0  o k = 0  o descending = FALSE  o threshold = 0  o msLevel = 1
```
I think you are missing a newline in the `cat()`:
```r
        for (i in seq_along(args)) {
            cat("  o ", names(args)[i], " = ", .cat_fun(args[[i]]), "\n", sep = "")
        }
```

to get 
```r
> x1@processingQueue
[[1]]
Object of class "ProcessingStep"
 Function: user-provided function
 Arguments:
  o intensity = user-provided function
  o msLevel = 1

[[2]]
Object of class "ProcessingStep"
 Function: user-provided function
 Arguments:
  o halfWindowSize = 2
  o method = MAD
  o snr = 0
  o k = 0
  o descending = FALSE
  o threshold = 0
  o msLevel = 1
```

File: R/ProcessingStep.R [hunk 2]
```diff
@@ -80,18 +80,23 @@ ProcessingStep <- function(FUN = character(), ARGS = list())  {
...
     cat("Object of class \"", class(object), "\"\n", sep = "")
-    cat(" Function: ", object@FUN, "\n", sep = "")
+    cat(" Function: ", .cat_fun(object@FUN), "\n", sep = "")
     args <- object@ARGS
     if (length(args) > 0) {
         cat(" Arguments:\n")
         for (i in seq_along(args)) {
-            cat("  o ", names(args)[i], " = ",
-                paste(args[[i]], collapse = ", "), "\n", sep = "")
+            cat("  o ", names(args)[i], " = ", .cat_fun(args[[i]]), sep = "")
```
Comment (lgatto) [inline R/ProcessingStep.R:99]: I think there's `\n` missing in your `cat()`

Comment (jorainer) [reply R/ProcessingStep.R:99]: yep, thanks. Will fix that - and add the correct unit tests.

**Comment by jorainer** [comment]
I fixed now the missing `"\n"` and added *real* unit tests for `show,ProcessingStep`.

**Review by lgatto — APPROVED** [review]

**Comment by jorainer** [comment]
I will also fix this in the release version

**Pull request merged**